activation/nipost.go

```diff
@@ -192,5 +175,135 @@ func (nb *NIPostBuilder) BuildNIPost(ctx context.Context, challenge *types.Hash3
+			log.String("poet_id", util.Bytes2Hex(poetServiceID)),
+			log.Stringer("challenge", *challenge),
+			log.Err(err))
+		return nil, &PoetSvcUnstableError{msg: "failed to submit challenge to poet service", source: err}
+	}
+
+	logger.With().Info("challenge submitted to poet proving service",
+		log.String("poet_id", util.Bytes2Hex(poetServiceID)),
+		log.String("round_id", round.ID),
+		log.Stringer("challenge", *challenge))
+
+	return &types.PoetRequest{
+		PoetRound:     round,
+		PoetServiceID: poetServiceID,
+	}, nil
+}
+
+// Submit the challenge to all registered PoETs.
+func (nb *NIPostBuilder) submitPoetChallenges(ctx context.Context, challenge *types.Hash32) []types.PoetRequest {
+	g, ctx := errgroup.WithContext(ctx)
+	poetRequestsChannel := make(chan types.PoetRequest, len(nb.poetProvers))
+	for _, poetProver := range nb.poetProvers {
+		poet := poetProver
+		g.Go(func() error {
+			if poetRequest, err := submitPoetChallenge(ctx, nb.log, poet, challenge); err == nil {
+				poetRequestsChannel <- *poetRequest
+			} else {
+				nb.log.With().Warning("failed to submit challenge to PoET", log.Err(err))
+			}
+			return nil
+		})
+	}
+	g.Wait()
+	close(poetRequestsChannel)
+
+	poetRequests := make([]types.PoetRequest, 0, len(nb.poetProvers))
+	for request := range poetRequestsChannel {
+		poetRequests = append(poetRequests, request)
+	}
+	return poetRequests
+}
+
+// awaitPoetProof concurrently waits for proofs from all PoET services that a challenge was submitted to
+// until it gets proofs from all PoETs or the context is canceled.
+// The best proof is returned or `nil` if proof was not received at all.
+func (nb *NIPostBuilder) awaitPoetProof(ctx context.Context, challenge *types.Hash32) types.PoetProofRef {
+	incomingPoetProofRefs := make(chan types.PoetProofRef, len(nb.state.PoetRequests))
+	g, ctx := errgroup.WithContext(ctx)
+
+	// Spawn workers to concurrently wait for PoET proofs
+	for _, poetService := range nb.state.PoetRequests {
+		svc := poetService
+		g.Go(func() error {
+			proofSubscription := nb.poetDB.SubscribeToProofRef(svc.PoetServiceID, svc.PoetRound.ID)
+			defer nb.poetDB.UnsubscribeFromProofRef(svc.PoetServiceID, svc.PoetRound.ID)
+			select {
+			case ref := <-proofSubscription:
+				nb.log.With().Debug("Worker got a new PoET proof",
+					log.String("poet_id", util.Bytes2Hex(svc.PoetServiceID)),
+					log.String("round_id", svc.PoetRound.ID),
+					log.Binary("ref", ref),
+				)
+				// We are interested only in proofs that we are members of
+				membership, err := nb.poetDB.GetMembershipMap(ref)
+				if err != nil {
+					nb.log.With().Panic("failed to fetch membership for poet proof",
+						log.Binary("challenge", challenge[:]))
+				}
+				if membership[*challenge] {
+					incomingPoetProofRefs <- ref
+				}
+			case <-ctx.Done():
+			}
+			return nil
+		})
+	}
+
+	result := make(chan types.PoetProofRef)
+
+	// Spawn consumer processing received proofs
+	go func() {
```

Comment:
I don't really follow what you mean here. Just using an `errgroup` doesn't guarantee that a goroutine will finish, even if the parent context of the errgroup is cancelled.

What if there is a "temporary" task to do, meaning spawn a few workers, and wait for them to finish, like in this case? If I use a "parent" errgroup I will not be able to `.Wait()` for them to finish.

The way to not use raw `go func() {}()` here would be to use a **separate** (different than `g`) and **local** errgroup. But what is the advantage? `errgroup` approach is just the same error-prone as the current one (one might forget to check if `ctx.Done()`. I think `errgroup` here is overkill and would complicate the code (need to select on two channels), though it would be more explicit. I'm sorry if I'm missing some obvious benefit here. 

Here I want to guarantee that all goroutines are finished before I return from this function. This goroutine is guaranteed to finish on app shutdown because `incomingPoetProofRefs` channel will be closed and I wait for its result.

Reply:
please see https://github.com/spacemeshos/go-spacemesh/pull/3603#discussion_r995305653

Reply:
but i am hoping all these spawn goroutines for processing proofs will be removed with the lazy approach to retrieve proofs from DB